cmd/zdb/zdb.c

```diff
@@ -9372,7 +9372,11 @@ main(int argc, char **argv)
 		if (argc != 3)
 			usage();
 		dump_opt['v'] = verbose;
-		error = dump_path(argv[0], argv[1], &object);
+		if (zdb_numeric(argv[1])) {
+			object = strtoull(argv[1], NULL, 0);
```

Comment:
There's a couple of problems here. `zdb_numeric()` and `strtoull(base=0)` permit different things, for example, `zdb_numeric` allows `abc`, but `strtoull()` does not, and will return 0.

The other more obvious problem is that this prevents a file name that `zdb_numeric()` returns true for, eg, there's no way to actually dump a file called `abc`.

Probably it needs a switch or magic value to say "interpret as number, not name". Maybe it's `-N` to match the "dataset is an objset id" case? Or maybe it can be a leading `/` on the filename, since that can't appear relative to the dataset? This seems to work reasonably well:

```c
diff --git cmd/zdb/zdb.c cmd/zdb/zdb.c
index 3b8b67bce..9018eea58 100644
--- cmd/zdb/zdb.c
+++ cmd/zdb/zdb.c
@@ -9372,8 +9372,12 @@ main(int argc, char **argv)
 		if (argc != 3)
 			usage();
 		dump_opt['v'] = verbose;
-		if (zdb_numeric(argv[1])) {
-			object = strtoull(argv[1], NULL, 0);
+		if (argv[1][0] == '/') {
+			errno = 0;
+			object = strtoull(&(argv[1])[1], NULL, 0);
+			error = errno;
+			if (error == 0 && object == 0)
+				error = EINVAL;
 		} else {
 			error = dump_path(argv[0], argv[1], &object);
 		}
```

Absent that, maybe trying a number if `dump_path()` returns `ENOENT`? It has the reverse problem in that you can't dump by object number if you have a file named that, so maybe it's not useful.